src/nemos/base_regressor.py

```diff
@@ -0,0 +1,465 @@
+import abc
+import inspect
+import warnings
+from typing import Any, NamedTuple, Optional, Tuple, Union, Dict
+
+import jax
+import jax.numpy as jnp
+import jaxopt
+from numpy.typing import ArrayLike, NDArray
+
+from . import utils, validation
+from ._regularizer_builder import AVAILABLE_REGULARIZERS, create_regularizer
+from .base_class import Base
+from .regularizer import Regularizer, UnRegularized
+from .typing import DESIGN_INPUT_TYPE, SolverInit, SolverRun, SolverUpdate
+
+
+class BaseRegressor(Base, abc.ABC):
+    """Abstract base class for GLM regression models.
+
+    This class encapsulates the common functionality for Generalized Linear Models (GLM)
+    regression models. It provides an abstraction for fitting the model, making predictions,
+    scoring the model, simulating responses, and preprocessing data. Concrete classes
+    are expected to provide specific implementations of the abstract methods defined here.
+    Below is a table listing the default and available solvers for each regularizer.
+
+    | Regularizer   | Default Solver   | Available Solvers                                           |
+    | ------------- | ---------------- | ----------------------------------------------------------- |
+    | UnRegularized | GradientDescent  | GradientDescent, BFGS, LBFGS, NonlinearCG, ProximalGradient, LBFGSB |
+    | Ridge         | GradientDescent  | GradientDescent, BFGS, LBFGS, NonlinearCG, ProximalGradient, LBFGSB |
+    | Lasso         | ProximalGradient | ProximalGradient                                            |
+    | GroupLasso    | ProximalGradient | ProximalGradient                                            |
+
+    Parameters
+    ----------
+    regularizer :
+        Regularization to use for model optimization. Defines the regularization scheme
+        and related parameters.
+        Default is UnRegularized regression.
+    regularizer_strength :
+        Float that is default None. Sets the regularizer strength. If a user does not pass a value, and it is needed for
+        regularization, a warning will be raised and the strength will default to 1.0.
+    solver_name :
+        Solver to use for model optimization. Defines the optimization scheme and related parameters.
+        The solver must be an appropriate match for the chosen regularizer.
+        Default is `None`. If no solver specified, one will be chosen based on the regularizer.
+        Please see table above for regularizer/optimizer pairings.
+    solver_kwargs :
+        Optional dictionary for keyword arguments that are passed to the solver when instantiated.
+        E.g. stepsize, acceleration, value_and_grad, etc.
+         See the jaxopt documentation for details on each solver's kwargs: https://jaxopt.github.io/stable/
+
+    See Also
+    --------
+    Concrete models:
+
+    - [`GLM`](../glm/#nemos.glm.GLM): A feed-forward GLM implementation.
+    - [`GLMRecurrent`](../glm/#nemos.glm.GLMRecurrent): A recurrent GLM implementation.
```

Comment:
This should be changed: leftover from past implementatoin
```suggestion
    - [`PopulationGLM`](../glm/#nemos.glm.PopulationGLM): A population GLM implementation.
```